src/apps/ENGINE/s_vartab.h

```diff
@@ -1,83 +1,59 @@
 #pragma once
 
 #include "data.h"
+#include "strutils.h"
+#include <unordered_map>
 #include <vector>
 
 // when segment_id is INVALID_SEGMENT_ID, variable segment is unloaded
 // and variable value and type undefined
 
-#define VAR_BUFFER_BLOCK_SIZE 1024
 #define INVALID_VAR_CODE 0xffffffff
 #define INVALID_SEGMENT_ID 0xffffffff
 
-struct VARINFO
+struct VarInfo
 {
-    VARINFO()
+    VarInfo();
+
+    bool IsArray()
     {
-        pDClass = nullptr;
-        name = nullptr;
-        type = UNKNOWN;
-        bArray = false;
-        elements = 0;
-        segment_id = INVALID_SEGMENT_ID;
-        hash = 0;
-    };
-    uint32_t segment_id;
-    uint32_t hash;
-    DATA *pDClass;
-    char *name;
-    uint32_t elements;
+        return elements > 1;
+    }
+
+    std::string name;
     S_TOKEN_TYPE type;
-    bool bArray;
+    std::shared_ptr<DATA> value;
```

Comment:
Is this shared_ptr used anywhere else? Or could it be a unique_ptr?

Reply:
With the current interface of VarTable it can't be replaced with a unique_ptr. VarTable::GetVar and VarTable::GetVarX require to make a copy of a VarInfo.

Reply:
I have not had a chance to review the code yet, but I believe we definitely shall try to avoid using std::shared_ptr there
It would be a significant drop in performance due to a control block
I will try to check this soon and propose a solution

Reply:
I rewrote the code, so it uses a const pointer to VarInfo instead of a copy. This allowed me to get rid of shared_ptr. Let me know if this is a good solution.